File: core/json_writer.py
import os
import json
import structlog
from typing import Any, Dict, List

logger = structlog.get_logger()

class JsonWriter:
    """
    Handles file writing operations for output results, change reports,
    and validation reports, ensuring proper directories exist.
    """
    
    @staticmethod
    def write_json(data: Any, filepath: str) -> None:
        """
        Writes data to a target JSON file, creating parent directories
        if they do not exist.
        """
        dir_name = os.path.dirname(filepath)
        if dir_name:
            os.makedirs(dir_name, exist_ok=True)
            
        try:
            with open(filepath, "w", encoding="utf-8") as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            logger.info("file_write_success", path=filepath)
        except Exception as e:
            logger.error("file_write_failed", path=filepath, error=str(e))
            raise

    @classmethod
    def generate_validation_report(cls, validation_records: Dict[str, List[str]], filepath: str) -> None:
        """
        Formats and writes a validation report listing warnings, errors,
        or data quality issues for each bank scraper.
        """
        report = {
            "validation_summary": {
                "total_banks_checked": len(validation_records),
                "banks_with_issues": sum(1 for errors in validation_records.values() if errors)
            },
            "failures_and_warnings": validation_records
        }
        cls.write_json(report, filepath)

File: core/jsonWriter.js
import fs from 'fs';
import path from 'path';
import { logger } from './logger.js';

export class JsonWriter {
  static writeJson(data, filepath) {
    const dirName = path.dirname(filepath);
    if (dirName) {
      fs.mkdirSync(dirName, { recursive: true });
    }

    try {
      fs.writeFileSync(filepath, JSON.stringify(data, null, 2), "utf-8");
      logger.info("file_write_success", { path: filepath });
    } catch (e) {
      logger.error("file_write_failed", { path: filepath, error: e.message });
      throw e;
    }
  }

  static generateValidationReport(validationRecords, filepath) {
    let issueBanksCount = 0;
    for (const errors of Object.values(validationRecords)) {
      if (errors && errors.length > 0) {
        issueBanksCount++;
      }
    }

    const report = {
      validation_summary: {
        total_banks_checked: Object.keys(validationRecords).length,
        banks_with_issues: issueBanksCount
      },
      failures_and_warnings: validationRecords
    };
    this.writeJson(report, filepath);
  }
}
